nomad/state/state_store_acl.go

```diff
@@ -235,3 +243,62 @@ func (s *StateStore) GetACLRoleByIDPrefix(ws memdb.WatchSet, idPrefix string) (m
 
 	return iter, nil
 }
+
+// fixTokenRoleLinks is a state helper that ensures the returned ACL token has
+// an accurate representation of ACL role links. The role links could have
+// become stale when a linked role was deleted or renamed. This will correct
+// them and generates a newly allocated token only when fixes are needed. If
+// the role links are still accurate, we just return the original token.
+func (s *StateStore) fixTokenRoleLinks(txn ReadTxn, original *structs.ACLToken) (*structs.ACLToken, error) {
+
+	// Track whether we have made an initial copy to ensure we are not
+	// operating on the token directly from state.
+	copied := false
```

Comment:
Just an idea - this would probably be easier to read if we always created the duplicate, then compare hashes at then end to decide if we should return the duplicate or the original